### PR TITLE
Extended keys serialization and path generation / derivation

### PR DESCRIPTION
**Pull request by A-Zak**
Hey Kevin,

I've added the extended keys serialization and also methods to generate key paths and to derive child keys with a path. I've also added some tests - for the serialization I've tested according to the BIP32 test vectors, and for the path, I've just tried to cover all edge cases...
I also tried to comply to your style :D

Cheers...

### CODE REVIEW

**Comment by kgreenek** [comment]
Wow thanks! I'll try to take a look at this tonight.

One piece of feedback I would give after taking a glance is to take a look
at the "params" pattern I've been starting to move to. For example see how
mainnetparameters conforms to the addressparameters protocol. This makes
things more flexible in the future, for example for alt coins.

**Comment by A-Zak** [comment]
I've just refactored that part erlier today :)

**Comment by kgreenek** [comment]
Sorry for the delay. Taking a look now. In general it looks good. Please watch your style - especially extra whitespace. I'm sorry there isn't better documentation; up until now it's been mostly just me working on this, so I'll try to point you to what I'm thinking.

File: BitcoinSwift/BitcoinMainNetParameters.swift [hunk 1]
```diff
@@ -49,4 +49,14 @@ public class BitcoinMainNetParameters: BitcoinParameters {
   public var blockChainStoreFileName: String {
     return "blockchain"
   }
+  
+  // MARK: - ExtendedKeyVersionParameters
+  
+  public var addresses: (pub:UInt32, prv: UInt32) {
```
Comment (kgreenek) [inline BitcoinSwift/BitcoinMainNetParameters.swift]: The name "addresses" is very unclear. These are not addresses. Also remember that the name will be part of the BitcoinMainNetParameters class (for example), so the names need to be unique. This is true of protocols in general. You should make the name sufficiently unique that somebody implementing the protocol won't have a name collision with another protocol implemented by the same class.

I recommend splitting this into two separate vars like:
extendedPublicKeyVersion & extendedPrivateKeyVersion

Also in general, names should be fully typed out. E.g. instead of "pub" just type out "publicKey". The whole rest of the code base is this way.

File: BitcoinSwift/ExtendedECKey.swift [hunk 2]
```diff
@@ -8,18 +8,144 @@
 
 import Foundation
 
+
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: extra whitespace.

File: BitcoinSwift/ExtendedECKey.swift [hunk 3]
```diff
@@ -8,18 +8,144 @@
...
+
+
+public enum KeyType {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: I don't think KeyType is really an accurate way to represent this. An ECKey can have either:
- both a private key and a public key
or
- just a public key

In the ECKey class, this is already represented by checking whether the private key is nil. So this is really a key pair, where the private key can be nil in some cases.

File: BitcoinSwift/ExtendedECKey.swift [hunk 4]
```diff
@@ -8,18 +8,144 @@
...
 public class ExtendedECKey : ECKey {
-
+  
   public let chainCode: SecureData
   public let index: UInt32
+  public let version: ExtendedKeyVersionParameters
+  public let parent: ExtendedECKey?
+  
+  
+  // This key's identifier (this corresponds exactly to a traditional bitcoin address)
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: comments should always end with a period (sorry, I know there isn't a published style guide. Please bear with all my nits.)

File: BitcoinSwift/ExtendedECKey.swift [hunk 5]
```diff
@@ -8,18 +8,144 @@
...
+  public var identifier:NSData {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: space after the colon

File: BitcoinSwift/ExtendedECKey.swift [hunk 6]
```diff
@@ -8,18 +8,144 @@
...
+    return publicKey.SHA256Hash().RIPEMD160Hash()
+  }
+  
+  // Return this key's fingerprint
+  public var fingerprint:NSData {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: here too. and elsewhere in this PR. There should be a space after the colon.

File: BitcoinSwift/ExtendedECKey.swift [hunk 7]
```diff
@@ -8,18 +8,144 @@
...
+  public let version: ExtendedKeyVersionParameters
+  public let parent: ExtendedECKey?
+  
+  
+  // This key's identifier (this corresponds exactly to a traditional bitcoin address)
+  public var identifier:NSData {
+    return publicKey.SHA256Hash().RIPEMD160Hash()
+  }
+  
+  // Return this key's fingerprint
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: Period after comments.

File: BitcoinSwift/ExtendedECKey.swift [hunk 8]
```diff
@@ -8,18 +8,144 @@
...
+    return self.identifier.subdataWithRange(NSMakeRange(0, 4))
+  }
+  
+  // Return the parent's fingerprint. 0x0000000 if master
+  public var parentFingerprint:NSData {
+    if let parent = parent {
+      return parent.fingerprint
+    }
+    else {
+      let masterprint: [UInt8] = [
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: Define masterFingerprint as a class var. Also, don't shorten names. Just call it masterFingerprint.

File: BitcoinSwift/ExtendedECKey.swift [hunk 9]
```diff
@@ -8,18 +8,144 @@
...
+        0x00, 0x00, 0x00, 0x00]
+      return NSData(bytes: masterprint, length: masterprint.count)
+    }
+  }
+  
+  // Find the depth of this key
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: Here and elsewhere. If you use "///" instead of just "//" when commenting functions and classes, then when the user control+clicks on the function, they'll see the documentation in the little bubble that pops up. Similar to Apple's framework classes.

Please do that on the other functions/vars in this PR too.

File: BitcoinSwift/ExtendedECKey.swift [hunk 10]
```diff
@@ -8,18 +8,144 @@
...
+      return parent.depth + 1
+    }
+    else {
+      return 0
+    }
+  }
+  
+  // True if this key has no parent
+  public var isMaster:Bool {
+    return self.parent == nil
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: no need for "self."

File: BitcoinSwift/ExtendedECKey.swift [hunk 11]
```diff
@@ -8,18 +8,144 @@
...
+  }
+  
+  // Find the master key by traversing the parents until a parentless key is found.
+  public var master:ExtendedECKey {
+    if let parent = self.parent {
+      return parent.master
+    }
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: the "else" should be on the same line as the "}". Elsewhere in this PR too.

File: BitcoinSwift/ExtendedECKey.swift [hunk 12]
```diff
@@ -8,18 +8,144 @@
...
+  }
+  
+  // Combines the path components into a string path separated with slashes (\)
+  public var path:String {
+    return "\\".join(self.pathComponents)
+  }
+  
+  // Serialize the extended key data
+  public func serializeExtendedKey(ofType type:KeyType = .PublicKey, version: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
+
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: extra whitespace.

File: BitcoinSwift/ExtendedECKey.swift [hunk 13]
```diff
@@ -8,18 +8,144 @@
...
+  // Combines the path components into a string path separated with slashes (\)
+  public var path:String {
+    return "\\".join(self.pathComponents)
+  }
+  
+  // Serialize the extended key data
+  public func serializeExtendedKey(ofType type:KeyType = .PublicKey, version: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
+    
+    func keyDataForType(type:KeyType) -> NSData {
+
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: extra whitespace here too.

File: BitcoinSwift/ExtendedECKey.swift [hunk 14]
```diff
@@ -8,18 +8,144 @@
...
+    }
+  }
+  
+  // Combines the path components into a string path separated with slashes (\)
+  public var path:String {
+    return "\\".join(self.pathComponents)
+  }
+  
+  // Serialize the extended key data
+  public func serializeExtendedKey(ofType type:KeyType = .PublicKey, version: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: Remove the "KeyType" enum and just have two separate functions "serializedExtendedPublicKey" and "serializedExtendedPrivateKey".

File: BitcoinSwift/ExtendedECKey.swift [hunk 15]
```diff
@@ -8,18 +8,144 @@
...
+    extKey.appendData(self.parentFingerprint)                                       // parent fingerprint
+    extKey.appendUInt32(self.index, endianness: .BigEndian)                         // child number
+    extKey.appendData(self.chainCode.mutableData)                                   // chain code
+    extKey.appendData(keyDataForType(type))                                         // public/private key
+    
+    return extKey
+  }
+  
+  // Encode the extended key into a base64check string
+  public func encodeExtendedKey(ofType type:KeyType = .PublicKey, version: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> String {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: Sorry, I know this isn't documented anywhere, the the maximum line length is 100 chars. Please wrap when longer than that.

File: BitcoinSwift/ExtendedECKey.swift [hunk 16]
```diff
@@ -81,13 +207,64 @@ public class ExtendedECKey : ECKey {
...
+    // guard agains overflows
+    if index > UInt32.max / 2 {
+      return nil
+    }
     return childKeyWithIndex(index + ExtendedECKey.hardenedIndexOffset())
   }
 
+  // Derive a key from the path.
+  // Supports both absolute or relative paths.
+  public func derive(path:String, isAbsolute isAbsolutePath:Bool = true) -> ExtendedECKey? {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: I would call this deriveFromExtendedKeyPath. Please be explicit with function/variable names.

File: BitcoinSwift/ExtendedECKey.swift [hunk 17]
```diff
@@ -81,13 +207,64 @@ public class ExtendedECKey : ECKey {
...
   public func childKeyWithHardenedIndex(index: UInt32) -> ExtendedECKey? {
+    // guard agains overflows
+    if index > UInt32.max / 2 {
+      return nil
+    }
     return childKeyWithIndex(index + ExtendedECKey.hardenedIndexOffset())
   }
 
+  // Derive a key from the path.
+  // Supports both absolute or relative paths.
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: Add comments here about the format of the extended key path.

File: BitcoinSwiftTests/ExtendedECKeyTests.swift [hunk 18]
```diff
@@ -287,4 +287,131 @@ class ExtendedECKeyTests: XCTestCase {
...
+    let mMaster = ExtendedECKey.masterKeyWithSeed(seed, version: BitcoinMainNetParameters.get())!
+    let tMaster = ExtendedECKey.masterKeyWithSeed(seed, version: BitcoinTestNetParameters.get())!
+    let mGrandChild = mMaster.childKeyWithIndex(0)!.childKeyWithHardenedIndex(0)!
+    let tGrandChild = tMaster.childKeyWithHardenedIndex(0)!.childKeyWithIndex(0)!
+    
+    XCTAssert(mGrandChild.version === BitcoinMainNetParameters.get())
+    XCTAssert(tGrandChild.version === BitcoinTestNetParameters.get())
+  }
+  
+  func testExtendedKeySerilization_TestVector1() {
```
Comment (kgreenek) [inline BitcoinSwiftTests/ExtendedECKeyTests.swift]: nit: No need for the underscore. The codebase doesn't do that anywhere else.

File: BitcoinSwiftTests/ExtendedECKeyTests.swift [hunk 19]
```diff
@@ -287,4 +287,131 @@ class ExtendedECKeyTests: XCTestCase {
...
+    // chain - m
+    XCTAssertEqual(key.encodeExtendedKey(ofType: .PublicKey),  "xpub661MyMwAqRbcFtXgS5sYJABqqG9YLmC4Q1Rdap9gSE8NqtwybGhePY2gZ29ESFjqJoCu1Rupje8YtGqsefD265TMg7usUDFdp6W1EGMcet8")
+    XCTAssertEqual(key.encodeExtendedKey(ofType: .PrivateKey), "xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi")
+    
+    // chain - m/0ʜ
+    key = key.childKeyWithHardenedIndex(0)
+    XCTAssertEqual(key.encodeExtendedKey(ofType: .PublicKey),  "xpub68Gmy5EdvgibQVfPdqkBBCHxA5htiqg55crXYuXoQRKfDBFA1WEjWgP6LHhwBZeNK1VTsfTFUHCdrfp1bgwQ9xv5ski8PX9rL2dZXvgGDnw")
+    XCTAssertEqual(key.encodeExtendedKey(ofType: .PrivateKey), "xprv9uHRZZhk6KAJC1avXpDAp4MDc3sQKNxDiPvvkX8Br5ngLNv1TxvUxt4cV1rGL5hj6KCesnDYUhd7oWgT11eZG7XnxHrnYeSvkzY7d2bhkJ7")
+    
+    // chain - m/0ʜ/1
```
Comment (kgreenek) [inline BitcoinSwiftTests/ExtendedECKeyTests.swift]: These tests are awesome, thank you!

File: BitcoinSwift/ExtendedECKey.swift [hunk 20]
```diff
@@ -36,7 +162,7 @@ public class ExtendedECKey : ECKey {
 
   /// Can return nil in the (very very very very) unlikely case the randomly generated private key
   /// is invalid. If nil is returned, retry.
-  public class func masterKeyWithSeed(seed: SecureData) -> ExtendedECKey? {
+  public class func masterKeyWithSeed(seed: SecureData, version:BitcoinParameters = BitcoinMainNetParameters.get()) -> ExtendedECKey? {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: Call this "params" instead of version.

**Comment by A-Zak** [comment]
Hey,

First of all, sorry it took me so long... We had some very tight schedules and I had very little time to spare...

About the comments, I can appreciate the style guide, and I've changed most of the things you've asked - all but two comments:
1) the max line length (100 chars) - I think it sometimes looks very bad when a line of code is broken down to several lines, especially when combined with the "no empty lines under any circumstance" style guide. So I left a couple of more than 100 char lines where I thought it looks better...
and 2) the default value for the params parameter in the initializers. I understand that creating a dependency to BitcoinMainNetParams makes it a tiny bit harder for someone to reuse this code for other alt coins, but I think that it's more than understandable that **Bitcoin**Swift will have a dependencies to Bitcoin classes, not to mention the great care you took to make it extensively 'protocoled' to make any such changes extremely easy...
In my opinion most of the code using BitcoinSwift will be used to on the main net, and with the default value in place, the benefits of a cleaner and more maintainable code, which is also less error prone, are far greater.

Let me know what you think... :)

**Comment by A-Zak** [comment]
P.S.
There's another pull request waiting with xCode 6.3 / Swift 1.2 compatibility fixes...

**Comment by kgreenek** [comment]
No worries. This is just a part time thing for me too :)

1) The char-limit is a hard limit. Please wrap everything so it's all under 100 chars. There is almost always a way to wrap it that looks nice, but in the rare cases when there isn't, it's still better to wrap manually than have your editor do the line-wrapping or scroll.

2) I'm worried more about people accidentally creating something for mainnet when they meant testnet and having a mismatch that breaks things in difficult ways. I always want to require params to be explicitly passed to it's clear what the intention is. Also, it should follow the pattern of passing params as the first argument.

Another thing, eventually I'm going to split up the codebase into libraries, and I don't want to introduce a dependency on the "BitcoinParams" library (or whatever it ends up being) when there shouldn't be. I'm a big supporter of keeping things 100% semantically pure and modular.

File: BitcoinSwift/ExtendedECKey.swift [hunk 21]
```diff
@@ -8,18 +8,151 @@
...
+  private class func masterParentFingerprint() -> NSData {
+    // TODO: Remove this once Swift supports class vars.
+    struct Static {
+      static let masterParentFingerprint: [UInt8] = [0x00, 0x00, 0x00, 0x00]
+    }
+    return NSData(bytes: Static.masterParentFingerprint,
+                 length: Static.masterParentFingerprint.count)
+  }
+  
+  private class func pathSeporator() -> Character {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: Spelling

File: BitcoinSwift/ExtendedECKey.swift [hunk 22]
```diff
@@ -8,18 +8,151 @@
...
+}
+
+
 /// An ExtendedECKey represents a key that is part of a DeterministicECKeyChain. It is just an
 /// ECKey except an additional chainCode parameter and an index are used to derive the key.
 /// Extended keys can be used to derive child keys.
 /// BIP 32: https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki
 public class ExtendedECKey : ECKey {
-
+
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: trailing whitespace

File: BitcoinSwift/BitcoinTestNetParamaters.swift [hunk 23]
```diff
@@ -49,4 +49,15 @@ public class BitcoinTestNetParameters: BitcoinParameters {
   public var blockChainStoreFileName: String {
     return "blockchain_testnet"
   }
+  
+  // MARK: - ExtendedKeyVersionParameters
+  
+  public var extendedPublicKeyVersion: NSData {
+    let publicKeyVersion: [UInt8] = [0x04, 0x35, 0x87, 0xcf]
+    return NSData(bytes: publicKeyVersion, length: publicKeyVersion.count)
+  }
```
Comment (kgreenek) [inline BitcoinSwift/BitcoinTestNetParamaters.swift]: nit: add a nl

File: BitcoinSwift/BitcoinMainNetParameters.swift [hunk 24]
```diff
@@ -49,4 +49,15 @@ public class BitcoinMainNetParameters: BitcoinParameters {
   public var blockChainStoreFileName: String {
     return "blockchain"
   }
+  
+  // MARK: - ExtendedKeyVersionParameters
+  
+  public var extendedPublicKeyVersion: NSData {
+    let publicKeyVersion: [UInt8] = [0x04, 0x88, 0xb2, 0x1e]
+    return NSData(bytes: publicKeyVersion, length: publicKeyVersion.count)
+  }
```
Comment (kgreenek) [inline BitcoinSwift/BitcoinMainNetParameters.swift]: nit: add a nl

File: BitcoinSwift/ExtendedECKey.swift [hunk 25]
```diff
@@ -8,18 +8,151 @@
 
 import Foundation
 
+public protocol ExtendedKeyVersionParameters: class {
+  
+  var extendedPublicKeyVersion: NSData { get }
+  var extendedPrivateKeyVersion: NSData { get }
+}
+
+
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: extra whitespace

File: BitcoinSwift/ExtendedECKey.swift [hunk 26]
```diff
@@ -8,18 +8,151 @@
...
+  /// return a serialized public key data.
+  public func serializedExtendedPublicKey(params: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
+    return serializedExtendedKey(version: params.extendedPublicKeyVersion, keyData: self.publicKey)
+  }
+  
+  /// return a serialized private key data.
+  public func serializedExtendedPrivateKey(params: ExtendedKeyVersionParameters = BitcoinMainNetParameters.get()) -> NSMutableData {
+    let keyData = NSMutableData()
+    keyData.appendUInt8(0x00)
+    keyData.appendData(self.privateKey.mutableData)
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: Whenever you are storing a private key in memory, you always need to be using SecureData.

File: BitcoinSwift/ExtendedECKey.swift [hunk 27]
```diff
@@ -81,13 +214,67 @@ public class ExtendedECKey : ECKey {
     childPrivateKey.appendSecureData(childPrivateKeyInt.secureData)
     assert(Int32(childPrivateKey.length) == ECKey.privateKeyLength())
     let childChainCode = indexHash[32..<64]
-    return ExtendedECKey(privateKey: childPrivateKey, chainCode: childChainCode, index: index)
+    return ExtendedECKey(privateKey: childPrivateKey, chainCode: childChainCode, index: index, parent:self)
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: nit: space after colon

File: BitcoinSwift/ExtendedECKey.swift [hunk 28]
```diff
@@ -8,18 +8,151 @@
...
+    }
+  }
+  
+  /// Returns the absolute path string.
+  public var path: String {
+    let seporator = String(ExtendedECKey.pathSeporator())
+    return seporator.join(self.pathComponents)
+  }
+  
+  private func serializedExtendedKey(#version: NSData, keyData: NSData) -> NSMutableData {
```
Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: What's the # in front of version?

Comment (kgreenek) [inline BitcoinSwift/ExtendedECKey.swift]: Also, this should be called serializedExtendedKeyWithVersion

**Comment by kgreenek** [comment]
Just following up with this. It's really close. I really do want this to be merged. I may pull it in and make the remaining changes myself too if you're busy :)
